cmd/action/migrate.go

```diff
@@ -0,0 +1,179 @@
+	}
+)
+
+func init() {
+	RootCmd.AddCommand(MigrateCmd)
+	MigrateCmd.AddCommand(MigrateDiffCmd)
+	MigrateCmd.PersistentFlags().StringVarP(&MigrateFlags.DirURL, migrateFlagDir, "", "file://migrations", "select migration directory using DSN format")
+	MigrateCmd.PersistentFlags().StringSliceVarP(&MigrateFlags.Schemas, migrateFlagSchema, "", nil, "set schema names")
+	MigrateCmd.PersistentFlags().BoolVarP(&MigrateFlags.Force, migrateFlagForce, "", false, "force a command to run on a broken migration directory state")
+	MigrateDiffCmd.Flags().StringVarP(&MigrateFlags.DevURL, migrateDiffFlagDevURL, "", "", "[driver://username:password@address/dbname?param=value] select a data source using the DSN format")
+	MigrateDiffCmd.Flags().StringVarP(&MigrateFlags.ToURL, migrateDiffFlagTo, "", "", "[driver://username:password@address/dbname?param=value] select a data source using the DSN format")
+	MigrateDiffCmd.Flags().SortFlags = false
+	cobra.CheckErr(MigrateDiffCmd.MarkFlagRequired(migrateDiffFlagDevURL))
+	cobra.CheckErr(MigrateDiffCmd.MarkFlagRequired(migrateDiffFlagTo))
+}
+
+func CmdMigrateDiffRun(cmd *cobra.Command, args []string) {
+	// Open a dev driver.
+	dev, err := DefaultMux.OpenAtlas(MigrateFlags.DevURL)
+	cobra.CheckErr(err)
+	// Open the migration directory. For now only local directories are supported.
+	dir, err := dir()
+	cobra.CheckErr(err)
+	// Get a state reader for the desired state.
+	desired, err := to(cmd.Context(), dev)
+	cobra.CheckErr(err)
+	// Only create one file per plan.
+	tfmt, err := migrate.NewTemplateFormatter(nameTmpl, contentTmpl)
+	cobra.CheckErr(err)
+	// Plan the changes and create a new migration file.
+	pl := migrate.NewPlanner(dev, dir, migrate.WithFormatter(tfmt))
+	var n string
+	if len(args) > 0 {
+		n = args[0]
+	}
+	plan, err := pl.Plan(cmd.Context(), n, desired)
+	cobra.CheckErr(err)
+	return
+	// Write the plan to a new file.
+	cobra.CheckErr(pl.WritePlan(plan))
+}
+
+// dir returns a migrate.Dir to use as migration directory. For now only local directories are supported.
+func dir() (migrate.Dir, error) {
+	parts := strings.SplitN(MigrateFlags.DirURL, "://", 2)
+	if len(parts) != 2 {
+		return nil, fmt.Errorf("invalid dsn %q", MigrateFlags.DirURL)
+	}
+	switch parts[0] {
+	case "file":
+		return migrate.NewLocalDir(parts[1])
+	default:
+		return nil, fmt.Errorf("unsupported driver %q", parts[0])
+	}
+}
+
+// to returns a migrate.StateReader for the given to flag.
+func to(ctx context.Context, d *Driver) (migrate.StateReader, error) {
+	parts := strings.SplitN(MigrateFlags.ToURL, "://", 2)
+	if len(parts) != 2 {
+		return nil, fmt.Errorf("invalid dsn %q", MigrateFlags.ToURL)
+	}
+	schemas := ApplyFlags.Schema
+	if n, err := SchemaNameFromURL(ctx, parts[0]); n != "" {
+		cobra.CheckErr(err)
+		schemas = append(schemas, n)
+	}
+	switch parts[0] {
+	case "file": // hcl file
+		f, err := ioutil.ReadFile(parts[1])
+		if err != nil {
+			return nil, err
+		}
+		realm := new(schema.Realm)
```

Comment:
```suggestion
		realm := &schema.Realm{}
```
Just to stay consistent, I try to not use `new` in the codebase, but it's just a suggestion to you.